packages/@aws-cdk/tmp-toolkit-helpers/src/api/diff/diff.ts

```diff
@@ -0,0 +1,296 @@
+ *
+ * Returns the diff if the changes are prompt-worthy, an empty object otherwise.
+ */
+export function formatSecurityDiff(
+  ioHelper: IoHelper,
+  oldTemplate: any,
+  newTemplate: cxapi.CloudFormationStackArtifact,
+  requireApproval: RequireApproval,
+  stackName?: string,
+  changeSet?: DescribeChangeSetOutput,
+): FormatSecurityDiffOutput {
+  const ioDefaultHelper = new IoDefaultMessages(ioHelper);
+
+  const diff = fullDiff(oldTemplate, newTemplate.template, changeSet);
+
+  if (diffRequiresApproval(diff, requireApproval)) {
+    ioDefaultHelper.info(format('Stack %s\n', chalk.bold(stackName)));
+
+    // eslint-disable-next-line max-len
+    ioDefaultHelper.warning(`This deployment will make potentially sensitive changes according to your current security approval level (--require-approval ${requireApproval}).`);
+    ioDefaultHelper.warning('Please confirm you intend to make the following modifications:\n');
+
+    // The security diff is formatted via `Formatter`, which takes in a stream
+    // and sends its output directly to that stream. To faciliate use of the
+    // global CliIoHost, we create our own stream to capture the output of
+    // `Formatter` and return the output as a string for the consumer of
+    // `formatSecurityDiff` to decide what to do with it.
+    const stream = new StringWriteStream();
+    try {
+      // formatSecurityChanges updates the stream with the formatted security diff
+      formatSecurityChanges(stream, diff, buildLogicalToPathMap(newTemplate));
+    } finally {
+      stream.end();
+    }
+    // store the stream containing a formatted stack diff
+    const formattedDiff = stream.toString();
+    return { formattedDiff };
+  }
+  return {};
+}
+
+/**
+ * Output of formatStackDiff
+ */
+export interface FormatStackDiffOutput {
+  /**
+   * Number of stacks with diff changes
+   */
+  readonly numStacksWithChanges: number;
+
+  /**
+   * Complete formatted diff
+   */
+  readonly formattedDiff: string;
+}
+
+/**
+ * Formats the differences between two template states and returns it as a string.
+ *
+ * @param oldTemplate the old/current state of the stack.
+ * @param newTemplate the new/target state of the stack.
+ * @param strict      do not filter out AWS::CDK::Metadata or Rules
+ * @param context     lines of context to use in arbitrary JSON diff
+ * @param quiet       silences \'There were no differences\' messages
+ *
+ * @returns the formatted diff, and the number of stacks in this stack tree that have differences, including the top-level root stack
+ */
+export function formatStackDiff(
+  ioHelper: IoHelper,
+  oldTemplate: any,
+  newTemplate: cxapi.CloudFormationStackArtifact,
+  strict: boolean,
+  context: number,
+  quiet: boolean,
+  stackName?: string,
+  changeSet?: DescribeChangeSetOutput,
+  isImport?: boolean,
+  nestedStackTemplates?: { [nestedStackLogicalId: string]: NestedStackTemplates }): FormatStackDiffOutput {
```

Comment:
i agree that this sucks and should be part of a class with `formatSecurityDiff`. but i don't wanna do it in this PR, it's already too complicated.